src/converters/tablecell-post-fixer.js

```diff
@@ -0,0 +1,138 @@
+ * from rendering in the editing pipeline - text nodes in single `<paragraph>` are rendered in the data pipeline as direct children
+ * of the `<td>` or `<th>` elements. In other cases `<paragraph>` elements are rendered as `<p>` blocks.
+ *
+ * To ensure proper mappings between model and view elements and positions in the editing pipeline the table feature will always render
+ * an element in the view: `<span>` for single or empty `<paragraph>` and `<p>` otherwise.
+ *
+ * Example:
+ *
+ *		<table>
+ *			<tableRow>
+ *				<tableCell><paragraph></paragraph></tableCell>
+ *				<tableCell><paragraph>foo</paragraph></tableCell>
+ *				<tableCell><paragraph baz="bar">foo</paragraph></tableCell>
+ *				<tableCell><heading2>bar</heading2><paragraph>baz</paragraph></tableCell>
+ *			</tableRow>
+ *		</table>
+ *
+ * The editor will render in the data pipeline:
+ *
+ *		<figure>
+ *			<table>
+ *				<tbody>
+ *					<tr>
+ *						<td></td>
+ *						<td>foo</td>
+ *						<td><p baz="bar">foo</p></td>
+ *						<td><h3>bar</h3><p>baz</p></td>
+ *					</tr>
+ *				</tbody>
+ *			</table>
+ *		</figure>
+ *
+ * and in the editing view (without widget markup):
+ *
+ *		<figure>
+ *			<table>
+ *				<tbody>
+ *					<tr>
+ *						<td><span></span></td>
+ *						<td><span>foo</span></td>
+ *						<td><p baz="bar">foo</p></td>
+ *						<td><h3>bar</h3><p>baz</p></td>
+ *					</tr>
+ *				</tbody>
+ *			</table>
+ *		</figure>
+ *
+ * @param {module:engine/model/model~Model} model
+ * @param {module:engine/controller/editingcontroller~EditingController} editing
+ */
+export default function injectTableCellPostFixer( model, editing ) {
+	editing.view.document.registerPostFixer( writer => tableCellPostFixer( writer, model, editing.mapper ) );
+}
+
+// The table cell post-fixer.
+//
+// @param {module:engine/view/writer~Writer} writer
+// @param {module:engine/model/model~Model} model
+// @param {module:engine/conversion/mapper~Mapper} mapper
+function tableCellPostFixer( writer, model, mapper ) {
+	const changes = model.document.differ.getChanges();
+
+	// While this is view post fixer only nodes that changed are worth investigating.
+	for ( const entry of changes ) {
+		// Attribute change - check if it is single paragraph inside table cell that has attributes changed.
+		if ( entry.type == 'attribute' && entry.range.start.parent.name == 'tableCell' ) {
+			const tableCell = entry.range.start.parent;
+
+			if ( tableCell.childCount === 1 ) {
+				const singleChild = tableCell.getChild( 0 );
+				const renameTo = Array.from( singleChild.getAttributes() ).length ? 'p' : 'span';
+
+				renameParagraphIfDifferent( singleChild, renameTo, writer, mapper );
+			}
+		} else {
+			// Check all nodes inside table cell on insert/remove operations (also other blocks).
+			const tableCell = entry.position && entry.position.parent;
+
+			if ( tableCell && tableCell.is( 'tableCell' ) ) {
```

Comment:
`tableCell.is( 'tableCell' )` 😉

Reply:
I could name it `possibleTableCell` ;P

Reply:
Or `parent`. But `tableCell` is fine too.